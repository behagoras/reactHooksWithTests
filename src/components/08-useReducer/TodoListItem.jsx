import React from 'react'

export default function TodoListItem({ todo = {}, editTodo, deleteTodo }) {

  return (
    <li key={todo.id} className="list-group-item">
      <input
        type="checkbox"
        name="done"
        id="done"
        checked={todo.done}
        className="form-check-input"
        onChange={(e) => editTodo({ id: todo.id, done: e.target.checked })}
      />
      <input
        type="text"
        name="todo"
        id="todo"
        value={todo.todo}
        className={`form-control todo ${todo.done ? 'complete' : ''}`}
        placeholder="Add your todo description"
        onChange={(e) => editTodo({ id: todo.id, todo: e.target.value })}
      />
      <div className="col-5">
        <button type="button" className="btn btn-outline-primary" onClick={() => deleteTodo(todo.id)}>-</button>
      </div>
    </li>
  )
}
