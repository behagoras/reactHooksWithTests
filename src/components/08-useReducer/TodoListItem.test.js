import { shallow } from 'enzyme'
import React from 'react'
import { todoFixture } from '../../fixtures/todoFixtures'
import TodoListItem from './TodoListItem'

const editTodo = jest.fn()
const deleteTodo = jest.fn()

describe('<TodoListItem />', () => {
  let component = shallow(<TodoListItem
    todo={todoFixture}
    editTodo={editTodo}
    deleteTodo={deleteTodo}
  />)

  beforeEach(() => {
    component = shallow(<TodoListItem
      todo={todoFixture}
      editTodo={editTodo}
      deleteTodo={deleteTodo}
    />)
  })

  it('should render correctly', () => {
    expect(component).toMatchSnapshot()
  })

  it('should call delete function ', () => {
    component.find('button').simulate('click')
    expect(deleteTodo).toHaveBeenCalledTimes(1)
    expect(deleteTodo).toHaveBeenCalledWith(todoFixture.id)
  })

  it('should call edit function ', () => {
    const value = 'new value'
    component
      .find('.todo')
      .simulate('change', { target: { value } })
    expect(editTodo).toHaveBeenCalledTimes(1)
    expect(editTodo).toHaveBeenCalledWith({
      id: todoFixture.id,
      todo: value,
    })
  })

  it('should render the correct input text', () => {
    expect(
      component
        .find('.todo')
        .prop('value'),
    )
      .toBe(todoFixture.todo)
  })

  it('should have the complete class if TODO.done', () => {
    const todo = { ...todoFixture, done: true }
    component = shallow(<TodoListItem
      todo={todo}
      editTodo={editTodo}
      deleteTodo={deleteTodo}
    />)
    expect(component.find('.todo').hasClass('complete')).toBeTruthy()
  })

  it('should not have the complete class if not TODO.done', () => {
    const todo = { ...todoFixture, done: false }
    component = shallow(<TodoListItem
      todo={todo}
      editTodo={editTodo}
      deleteTodo={deleteTodo}
    />)
    expect(component.find('.todo').hasClass('complete')).toBeFalsy()
  })
})
