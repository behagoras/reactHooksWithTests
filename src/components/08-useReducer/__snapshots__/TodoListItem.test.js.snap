// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListItem /> should render correctly 1`] = `
<li
  className="list-group-item"
  key="123"
>
  <input
    checked={true}
    className="form-check-input"
    id="done"
    name="done"
    onChange={[Function]}
    type="checkbox"
  />
  <input
    className="form-control todo complete"
    id="todo"
    name="todo"
    onChange={[Function]}
    placeholder="Add your todo description"
    type="text"
    value="read a book"
  />
  <div
    className="col-5"
  >
    <button
      className="btn btn-outline-primary"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
  </div>
</li>
`;
